### PR TITLE
Action buffer

### PR DESCRIPTION
**Pull request by andrewcoh**
### Proposed change(s)

Describe the changes made in this PR.

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:63]:
```suggestion
    def from_dict(buff: Dict[str, np.ndarray]) -> "AgentAction":
```

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:79]: what is `all_discrete_list` for?

Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/utils.py:79]: SAC uses all_log_probs for discrete actions instead of just the log probs of the actions taken.

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (ervteng) [inline ml-agents-envs/mlagents_envs/base_env.py:258]: Default should be `[]`

File: ml-agents-envs/mlagents_envs/environment.py
Comment (andrewcoh) [inline ml-agents-envs/mlagents_envs/environment.py:349]: Still not addressed.

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:118]: Why is this one a List? Can the tensors here he more than single value?

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:119]: I think we should be able to make a method called something like `get_probs_of_action` that would take as input a discrete action and compute the probability of that action given `all_discrete_list`.

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:114]: Let's try to make this more explicit : 
```suggestion
    :param discrete_list: List of Torch tensors each corresponding to log probs of the discrete actions that were sampled.
    :param all_discrete_list: List of Torch tensors each corresponding to all log probs of
    a discrete action branch, even the discrete actions that were not sampled. all_discrete_list is a list of Tensors,
    each Tensor corresponds to one discrete branch log probabilities. 
```

File: ml-agents/mlagents/trainers/ppo/optimizer_tf.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/ppo/optimizer_tf.py:315]:
```suggestion
```

File: ml-agents/mlagents/trainers/tf/components/bc/module.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/tf/components/bc/module.py:121]: else raise("Not implemented yet for Hybrid Actions")

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/utils.py:118]: the multicategorical distribution produces a list of categorical distributions, one for each branch.  it makes computing log probs simpler etc.

File: ml-agents/mlagents/trainers/ppo/optimizer_tf.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/ppo/optimizer_tf.py:315]: ive been caught redhanded

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/utils.py:119]: thats roughly what distribution.log_prob does and how it is used.

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (ervteng) [reply ml-agents-envs/mlagents_envs/base_env.py:247]: I vote for something simple. Maybe `ActionTuple`, `ActionSteps` or just `Actions`

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (ervteng) [reply ml-agents-envs/mlagents_envs/base_env.py:342]: Might be beneficial for people debugging their custom trainers/envs, and seems a logical complement to `empty_action`. Am OK with it moved too though.

File: ml-agents/mlagents/trainers/trajectory.py
Comment (ervteng) [inline ml-agents/mlagents/trainers/trajectory.py:116]: It seems like this relies on the fact that discrete actions are last in the dict. I'd make it explicit so that if the order of the dict changes it won't break.

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/utils.py]: log_probs is the probability of a single action, all_log_probs is all the probs of actions on a given branch

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (andrewcoh) [reply ml-agents-envs/mlagents_envs/base_env.py:247]: ActionTuple it is

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (andrewcoh) [reply ml-agents-envs/mlagents_envs/base_env.py:342]: This is used by the tests only but was part of the behavior spec which is why it has made its way into the action spec. I can remove it if either of you have strong feelings

Comment (andrewcoh) [reply ml-agents-envs/mlagents_envs/base_env.py:342]: I think you two are talking about different functions. Vince are you referring to `_validate_action` or `random_action`?

File: ml-agents/mlagents/trainers/tf/components/bc/module.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/tf/components/bc/module.py:121]: Can we hold off on this since its not going into master until the ActionModel is merged?

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (andrewcoh) [reply ml-agents-envs/mlagents_envs/base_env.py:307]: sounds good to me

File: ml-agents-envs/mlagents_envs/environment.py
Comment (andrewcoh) [reply ml-agents-envs/mlagents_envs/environment.py:422]: Meant as a TODO for the C# changes that change the proto to accept both continuous and discrete. Poorly worded on my part.

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (andrewcoh) [reply ml-agents-envs/mlagents_envs/base_env.py]: done

File: ml-agents/mlagents/trainers/env_manager.py
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/env_manager.py:150]: bump

File: ml-agents/mlagents/trainers/demo_loader.py
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/demo_loader.py:76]: bump

File: ml-agents-envs/mlagents_envs/tests/test_envs.py
Comment (vincentpierre) [reply ml-agents-envs/mlagents_envs/tests/test_envs.py:104]: bump

File: ml-agents/mlagents/trainers/policy/tf_policy.py
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/policy/tf_policy.py:276]: bump

Comment (andrewcoh) [reply ml-agents/mlagents/trainers/policy/tf_policy.py:276]: It is split in torch policy (via the `to_numpy_dict` utility on ActionLogProbs). This is to maintain compatibility with TF and the changes to how things are added to the buffer.

File: ml-agents/mlagents/trainers/env_manager.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/env_manager.py:150]: I made it a static method on the env manager since only the env_managers use it. We don't have a general utils file though which may be a better place for this to live

File: ml-agents/mlagents/trainers/demo_loader.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/demo_loader.py:76]: This is changed in the ActionModel PR with the TODO pointing to the proto changes.  Since this isn't going into master until its all together, can we hold off on the error?

File: ml-agents/mlagents/trainers/policy/policy.py
Comment (chriselion) [inline ml-agents/mlagents/trainers/policy/policy.py:52]: Why not `Dict[str, ActionTuple]`? a dictionary with two hardcoded keys (in `make_empty_previous_action()`) doesn't make much sense.

File: ml-agents/mlagents/trainers/policy/torch_policy.py
Comment (chriselion) [inline ml-agents/mlagents/trainers/policy/torch_policy.py:204]: Won't this convert tensor->nparray twice? (I guess the old code did this too.)

Comment (chriselion) [reply ml-agents/mlagents/trainers/policy/torch_policy.py:204]: Something like
```python
action_dict = action.to_numpy_dict()
run_out["action"] = action_dict
run_out["pre_action"] = action_dict["continuous_action"] if if self.use_continuous_act else None
```

File: ml-agents/mlagents/trainers/ppo/optimizer_tf.py
Comment (chriselion) [inline ml-agents/mlagents/trainers/ppo/optimizer_tf.py:341]: Should that comment be a TODO?

File: ml-agents/mlagents/trainers/policy/tf_policy.py
Comment (chriselion) [inline ml-agents/mlagents/trainers/policy/tf_policy.py:275]: I know it's not new to this PR, but should we move these strings to constant/enums? "pre_action" and "actions_pre" has me scared, and "continuous" is generally pretty typo-prone.

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (andrewcoh) [reply ml-agents-envs/mlagents_envs/base_env.py:257]: This was a check that was in `set_actions` from a long while ago that got moved to validate_actions in the action spec PR i.e. https://github.com/Unity-Technologies/ml-agents/pull/4586/files#diff-9cbe0b3e0de12842ef3b86098623226db8a4599cddfd86c1ab3f6a2a09ae7d06L372. I can remove it in this PR but it was not something I introduced myself.

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (andrewcoh) [reply ml-agents-envs/mlagents_envs/base_env.py:350]: I'm ok with moving it but this is a holdover from the ActionSpec PR (in which you suggested to make it private).

File: ml-agents/mlagents/trainers/policy/policy.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/policy/policy.py]: Good call out. I was doing this because it made writing to the buffer easier (e.g. using the string) but in reality we only ever use the previous action if its discrete. I've updated the PR to only save the discrete actions.

File: ml-agents/mlagents/trainers/ppo/optimizer_tf.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/ppo/optimizer_tf.py:341]: Not a TODO, just pointing out the fact that im getting continuous/discrete explicitly due to the buffer changes to support hybrid actions

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (vincentpierre) [reply ml-agents-envs/mlagents_envs/base_env.py:354]: I have multiple personalities when I review

File: ml-agents/mlagents/trainers/demo_loader.py
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/demo_loader.py:76]: Sounds good

File: ml-agents/mlagents/trainers/policy/tf_policy.py
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/policy/tf_policy.py:276]: Ah, I forgot about TF compatibility. My bad.

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (vincentpierre) [inline ml-agents-envs/mlagents_envs/base_env.py:258]: I think I would prefer having this not be Optional. When the action is discrete only, the continuous action becomes of shape (n_agents, 0). I think it is slightly more general than having a None if there is no action because this way, we can eventually consider all actions to be hybrid, only some of them happen to have 0 discrete or continuous actions. @chriselion what do you think? I might be missing something

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:18]: I know this comes a bit late, but since this class is pretty large, maybe we should have a `agent_action.py` for it?

Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/utils.py:18]: How about `action_utils.py` and we can include `ActionLogProbs` as well. Or would you prefer each have their own file?

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (chriselion) [reply ml-agents-envs/mlagents_envs/base_env.py:258]: I think for initializer arguments, keeping them Optional is nice. If you want to ensure that the fields themselves are non-None, you could "allocate" the np array on construction if they weren't provided (but not sure that's meaningful since there would be 0 elements).

As a side note, I think we're assuming that `continuous.shape[0] == discrete.shape[0]` - might be good to check that too.

File: ml-agents/mlagents/trainers/ppo/optimizer_tf.py
Comment (chriselion) [reply ml-agents/mlagents/trainers/ppo/optimizer_tf.py:341]: Ah, forgot this was TF code, so the implicit TODO is "burn it all"

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (andrewcoh) [reply ml-agents-envs/mlagents_envs/base_env.py:258]: We do check that shape dimensions in `_validate_action` on the ActionSpec. It gets called in set_actions/set_actions_for_agent in the LLAPI

File: ml-agents/mlagents/trainers/torch/utils.py
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/utils.py:18]: I would prefer each their own files

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (ervteng) [reply ml-agents-envs/mlagents_envs/base_env.py:256]: Why can't the default be `None` but in the constructor assigns an empty array when None is specified? This is a common pattern for mutable default parameters

File: ml-agents-envs/mlagents_envs/environment.py
Comment (ervteng) [inline ml-agents-envs/mlagents_envs/environment.py:425]:
```suggestion
                # TODO: This check will be removed when the proto supports hybrid actions
```

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (andrewcoh) [reply ml-agents-envs/mlagents_envs/base_env.py]: `_validate_actions` will check this`

File: ml-agents/mlagents/trainers/trajectory.py
Comment (ervteng) [inline ml-agents/mlagents/trainers/trajectory.py:11]: Can these be the same datatype of the action? E.g. `ActionTuple`

Comment (andrewcoh) [reply ml-agents/mlagents/trainers/trajectory.py:11]: https://github.com/Unity-Technologies/ml-agents/pull/4651

File: ml-agents/mlagents/trainers/env_manager.py
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/env_manager.py:151]: @vincentpierre this is removed in the ActionModel

**Review by ervteng — APPROVED** [review]
I think the remaining issues will end up being fixed by #4651, this part LGTM

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (vincentpierre) [inline ml-agents-envs/mlagents_envs/base_env.py:369]: why check if self.continuous_size > 0 ? same question line 36 with discrete.

File: ml-agents-envs/mlagents_envs/environment.py
Comment (vincentpierre) [inline ml-agents-envs/mlagents_envs/environment.py:426]: why check on the shape[1] rather than action_specs.continuous_size?

**Review by vincentpierre — APPROVED** [review]

File: ml-agents-envs/mlagents_envs/base_env.py
Comment (andrewcoh) [reply ml-agents-envs/mlagents_envs/base_env.py:369]: This is removed in the ActionModel because the defaults make it unnecessary

File: ml-agents-envs/mlagents_envs/environment.py
Comment (andrewcoh) [reply ml-agents-envs/mlagents_envs/environment.py:426]: ActionSpec is not available here. This is the only way I think we can determine if the action space is continuous or discrete in this function. **Critically, this is just to support the old proto until the new proto is merged. This is just temporary so that the communication protocol works until the new protos are in place.**

**Pull request merged**
